stout/flags.cc

```diff
@@ -18,35 +18,102 @@ namespace stout::flags {
 
 ////////////////////////////////////////////////////////////////////////
 
+std::vector<std::unique_ptr<Parser>> Parser::nested_parsers;
+std::map<std::string, google::protobuf::Message*> Parser::subcommands;
+
+////////////////////////////////////////////////////////////////////////
+
 void Parser::AddAllOrExit(google::protobuf::Message* message) {
   const auto* descriptor = message->GetDescriptor();
 
+  auto exit_on_missing_name =
+      [](const google::protobuf::FieldDescriptor* field) {
+        std::cerr
+            << "Missing at least one flag name in 'names' for field '"
+            << field->full_name() << "'"
+            << std::endl;
+        std::exit(1);
+      };
+
+  auto exit_on_missing_help =
+      [](const google::protobuf::FieldDescriptor* field) {
+        std::cerr
+            << "Missing flag 'help' for field '"
+            << field->full_name() << "'"
+            << std::endl;
+        std::exit(1);
+      };
+
   for (int i = 0; i < descriptor->field_count(); i++) {
     const auto* field = descriptor->field(i);
 
-    const auto& flag = field->options().GetExtension(stout::v1::flag);
+    // We need this descriptor for the subcommand's logic.
+    const google::protobuf::OneofDescriptor* oneof =
+        field->real_containing_oneof();
+
+    // Check if the current field of the message is in 'oneof'.
+    if (oneof != nullptr) {
+      // Check if the name of 'oneof' is 'subcommand'. If no -> exit with
+      // an error.
+      if (oneof->name() != "subcommand") {
+        std::cerr << "'oneof' field must have 'subcommand' name. "
+                  << "Other names are illegal"
+                  << std::endl;
+        exit(1);
+      } else {
+        // Subcommands must have stout.v1.subcommand option.
+        if (!field->options().HasExtension(stout::v1::subcommand)) {
+          std::cerr << "Every field of the 'oneof subcommand' must "
+                       "be annotated with a stout.v1.subcommand option"
+                    << std::endl;
+          exit(1);
+        } else {
+          // Check for missing 'names' and 'help'.
+          const auto& subcommand =
+              field->options().GetExtension(stout::v1::subcommand);
+
+          if (subcommand.names().empty()) {
+            exit_on_missing_name(field);
+          }
+
+          if (subcommand.help().empty()) {
+            exit_on_missing_help(field);
+          }
+
+          for (const auto& name : subcommand.names()) {
+            AddOrExit(name, field, message);
```

Comment:
This doesn't make any sense ... you're adding something for a flag, but this is a subcommand. the `AddOrExit()` function updates the `fields_` member, but this PR never actually uses the `fields_` member, so why did it need to get updated?

Instead, this is where you should be creating and saving the nested parser!